llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -3305,6 +3387,11 @@ RISCVTargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *AI) const {
   if (AI->isFloatingPointOperation())
     return AtomicExpansionKind::CmpXChg;
 
+  // The atomic RMW instructions are only available in capmode, so for the
+  // hybrid ABI we have expand these operations using a LR/SC sequence.
+  if (needsExplicitAddressingModeAtomics(AI, Subtarget))
+    return AtomicExpansionKind::LLSC;
```

Comment:
This still is differing. It should be doing things _exactly the same_ as the non-explicit versions, just with different expansions in ExpandAtomicPseudoInsts. No new intrinsics should be needed, just a bunch of SelectionDAG pseudos and patterns mirroring what's already there.

Reply:
Seems like quite a lot of effort. Maybe I'll get to it next year

Reply:
FWIW `AtomicExpansionKind::LLSC` is particularly broken, as it's done far too early in CodeGen so spills and reloads can sneak in between the two, which is highly problematic on architectures that permit very simple implementations with a single link/reserved bit. RISC-V's forward-progress guarantee requires the code to not use any memory instructions in between the LR and the SC, among other things. This is why the pseudos exist and are expanded so late, otherwise RISC-V would just use the existing expansion code and save a lot of effort.

Reply:
Good to know, I assumed it would work considering that ARM/AArch64 uses it.

Reply:
The Arm ARM probably mandates a less-stupid implementation, although technically maybe if you put an atomic on the stack you can get issues there, but why on earth would you do that.

Reply:
(See https://bugs.llvm.org/show_bug.cgi?id=32020 for a real-world case of this mattering)